Excel import 2

RomanBoyars opened the pull request.

RomanBoyars committed:
Импорт из excel

RomanBoyars committed:
забытая либа

RomanBoyars committed:
WIP мпорт CSV
Что исправлено:
Опечатки
Работа с VK API
Статистика по импорту
Что осталось:
Поддержка формата CSV

RomanBoyars committed:
Добавлен столбец с балансом

RomanBoyars committed:
Merge branch 'excel_import'

# Conflicts:
#	crm/templatetags/html_helper.py
#	crm/views/manager/client.py

RomanBoyars committed:
Поправлены замечания

RomanBoyars committed:
vk fix

ciklop commented on crm/views/manager/client.py:
```diff
@@ -332,8 +375,10 @@ def form_valid(self, form):
+            for index in get_duplicate_indeces(vk_user_ids, vk_user_id):
+                clients_to_add[index].vk_user_id = vk_user_id
         Client.objects.bulk_create(clients_to_add)
+        self.request.session['import_errors'] = errors
```
если ты это делаешь в одном view - то тебе сессия не нужна. запиши errors в self, и переопредели get_context_data в котором будешь проверять - есть ли в self errors, и если есть - добавлять в контекст.

RomanBoyars replied:
так это разные view

ciklop replied:
Тогда надо чистить данные из сесии после чтения либо не давать просто так заходить на страницу. 
Суть в том что на эту страницу можно попасть прямым образом, и в сесии будут висеть устаревшие данные. Такое лучше подчищать

ciklop commented on crm/templatetags/html_helper.py:
```diff
@@ -100,12 +102,36 @@ def get_vk_user_ids(vk_user_domains):
 
     return vk_user_ids
 
+def get_duplicate_indeces(seq,item):
```
это всё лучше переписать на такой вариант
```python
return [index for index, value in enumerate(seq) if value == item]
```

RomanBoyars replied:
такой вариант будет дольше выполняться
https://stackoverflow.com/questions/5419204/index-of-duplicates-items-in-a-python-list/23294343

ciklop replied:
Да, тут согласен, этот вариант быстрее.

ciklop commented on crm/templates/crm/manager/client/import_report.html:
```diff
@@ -0,0 +1,47 @@
+            </tr>
+          </thead>
+          <tbody>
+          {% for k, v in request.session.import_errors.items %}
```
надо переделать на вариант с контекстом

RomanBoyars committed:
Поправлено оформление кода
Русский текст об ошибке в функции парсинга даты
Поправлены ошибки в именовании ф-ции по поиску ключей дубликатов в массиве

RomanBoyars committed:
Правки по CR
Данные из сессии переносятся в контекст представления ImportReport
Затем удаляются из сессии в методе get

RomanBoyars committed:
Замена ошибки 404

RomanBoyars committed:
Поправил проверку данных в сессии

INovozhilov commented on crm/templatetags/html_helper.py:
```diff
@@ -100,13 +102,38 @@ def get_vk_user_ids(vk_user_domains):
         user_id = 'id' + str(result['id'])
         screen_name = result['screen_name']
         if screen_name in vk_user_domains:
```
что это за извращение? Цикл внутри цикла. Напиши обычный один цикл по vk_user_domains с одним IF внутри.
Будет работать быстро, и будет легко читаться.

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -293,18 +312,54 @@ def form_valid(self, form):
-            birthday = row[cell.column_index_from_string(birthday_col) - 1].value
             try:
-                balance = row[cell.column_index_from_string(balance_col) - 1].value
+                phone_raw = row[cell.column_index_from_string(phone_col) - 1].value
```
а phone_col может быть не задан? Например если пользователь не хочет импортировать телефон (нет его).

RomanBoyars replied:
Ну это не предусмотрено, это тогда лучше сделать галочку на всех опциональных столбцах. В задаче не было сказано ничего про опциональные столбцы.

RomanBoyars replied:
Там форма не позволит отправить с пустым полем

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -293,18 +312,54 @@ def form_valid(self, form):
+
+        for index, row in enumerate(iter_rows):
+            try:
+                name = row[cell.column_index_from_string(name_col) - 1].value
```
column_index_from_string -1 - сохранил бы уже в переменные и не дергал каждый раз в цикле.

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -293,18 +312,54 @@ def form_valid(self, form):
             try:
-                balance = row[cell.column_index_from_string(balance_col) - 1].value
+                phone_raw = row[cell.column_index_from_string(phone_col) - 1].value
+                phone = re.sub("\D", "", str(phone_raw))
```
Лучше бы в отдельную функцию вынести, для облегчения читаемости и упрощения дальнейшей доработки импорта.

RomanBoyars replied:
что я должен вынести в отдельные функции? Блоки try-except, там они завязаны на цикл, их не получится вынести.

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -293,18 +312,54 @@ def form_valid(self, form):
+
+
+            try:
+                balance = float(row[cell.column_index_from_string(balance_col) - 1].value.replace(',','.'))
```
в отдельную функцию

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -293,18 +312,54 @@ def form_valid(self, form):
+            except IndexError:
+                birthday = None
+            except ValueError:
+                errors["{}{}".format(birthday_col, index+1)] = "Неверный формат даты. Допустимые форматы: ГГГГ-ММ-ДД, ДД.ММ.ГГГГ, ДД/ММ/ГГГГ, ДД-ММ-ГГГГ."
```
Текст с перечислением допустимх форматов должен жить недалеко от самих форматов. Иначе при изменении будет сложно синхронизировать

RomanBoyars replied:
текст с перечислением форматов недалеко от самих форматов это как? у меня функция парсинга даты в html_helper, я тогда туда напишу ALLOWED_DATE_FORMATS константу?

INovozhilov commented on crm/views/manager/client.py:
```diff
@@ -332,8 +388,10 @@ def form_valid(self, form):
 
         vk_user_ids = get_vk_user_ids(vk_domains)
         for vk_user_id in vk_user_ids:
```
Цикл в цикле где он вроде как не нужен.

RomanBoyars committed:
Правки по CR

INovozhilov committed:
Merge branch 'master' into excel_import_2
